lib/types.nix

```diff
@@ -613,6 +618,95 @@ rec {
       nestedTypes.elemType = elemType;
     };
 
+    attrTag = tags: attrTagWith { inherit tags; };
+
+    attrTagWith = args@{ tags }:
+      let
+        tags =
+          mapAttrs
+            (n: opt:
+              builtins.addErrorContext "while checking that attrTag tag ${lib.strings.escapeNixIdentifier n} is an option with a type${inAttrPosSuffix args.tags n}" (
+                assert opt._type == "option";
+                opt // {
+                  declarations = opt.declarations or (
+                    let pos = builtins.unsafeGetAttrPos n args.tags;
+                    in if pos == null then [] else [ pos.file ]
+                  );
+                  declarationPositions = opt.declarationPositions or (
+                    let pos = builtins.unsafeGetAttrPos n args.tags;
+                    in if pos == null then [] else [ pos ]
+                  );
+                }
+              ))
+            args.tags;
+        choicesStr = concatMapStringsSep ", " lib.strings.escapeNixIdentifier (attrNames tags);
+      in
+      mkOptionType {
+        name = "attrTag";
+        description = "attribute-tagged union";
+        getSubOptions = prefix:
+          mapAttrs
+            (tagName: tagOption: {
+              "${lib.showOption prefix}" =
+                tagOption // {
+                  loc = prefix ++ [ tagName ];
+                  definitions = [];
+                };
+            })
+            tags;
+        substSubModules = m: attrTagWith { tags = mapAttrs (n: opt: opt // { type = (opt.type or types.unspecified).substSubModules m; }) tags; };
```

Comment:
The `types.unspecified` looks odd here, is that necessary or can we avoid it?

Reply:
In a module, the presence of `opt.type` would be guaranteed by `fixupOptionType`.
Calling that seems heavy-handed considering that this works, and is basically free, in terms of eval performance.

Reply:
I've removed `substSubModules` because without `getSubmodules`, this turns out to be dead code anyway.
So I've found while trying to test it.

I don't think `getSubmodules` is feasible to implement, because we don't have a single set of modules to return, let alone put back with `substSubModules`

I'll be happy to do something about this later if we find some sub-optimal generated docs.